### PR TITLE
[SDPA] relax restriction on L from query and attention_mask being equal

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - *relax restriction on L from query and attention_mask being equal*

### Tickets:
 - *CVS-129000*

### CODE REVIEW

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 1]
```diff
@@ -74,8 +74,8 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
         if (attention_mask_rank.is_static() && attention_mask_rank != 0) {
             const auto& attention_mask_rank_len = attention_mask_rank.get_length();
             bool attention_mask_input_correctness = attention_mask_rank_len >= 2 &&
-                                                    DimType::merge(l_dim, l_dim, *(attention_mask.end() - 2)) &&
-                                                    DimType::merge(s_dim, s_dim, *(attention_mask.end() - 1));
+                                                    DimType::broadcast_merge(l_dim, l_dim, *(attention_mask.end() - 2)) &&
+                                                    DimType::broadcast_merge(s_dim, s_dim, *(attention_mask.end() - 1));
```
Comment (jane-intel) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:78]:
```suggestion
                attention_mask_rank_len >= 2 && DimType::broadcast_merge(l_dim, l_dim, *(attention_mask.end() - 2)) &&
                DimType::broadcast_merge(s_dim, s_dim, *(attention_mask.end() - 1));
```

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 2]
```diff
@@ -74,8 +74,8 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
         if (attention_mask_rank.is_static() && attention_mask_rank != 0) {
             const auto& attention_mask_rank_len = attention_mask_rank.get_length();
             bool attention_mask_input_correctness = attention_mask_rank_len >= 2 &&
```
Comment (jane-intel) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:76]:
```suggestion
            bool attention_mask_input_correctness =
```

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 3]
```diff
@@ -73,9 +73,9 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
         const auto& attention_mask_rank = attention_mask.rank();
         if (attention_mask_rank.is_static() && attention_mask_rank != 0) {
             const auto& attention_mask_rank_len = attention_mask_rank.get_length();
-            bool attention_mask_input_correctness = attention_mask_rank_len >= 2 &&
-                                                    DimType::merge(l_dim, l_dim, *(attention_mask.end() - 2)) &&
-                                                    DimType::merge(s_dim, s_dim, *(attention_mask.end() - 1));
+            bool attention_mask_input_correctness =
+                attention_mask_rank_len >= 2 && DimType::broadcast_merge(l_dim, l_dim, *(attention_mask.end() - 2)) &&
+                DimType::broadcast_merge(s_dim, s_dim, *(attention_mask.end() - 1));
```
Comment (mitruska) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:78]: Would be great to cover this change with some tests, at least by "type_prop" for the op creation. 
https://github.com/openvinotoolkit/openvino/blob/eededc2a3a55a3c45a2e8053313d031bb35ce445/src/core/tests/type_prop/scaled_dot_product_attention.cpp#L250-L260
It should be also verified that plugins are able to broadcast the `attention_mask`, and cover such scenario in the plugins tests. Let's create a ticket for the tests if the fix is urgent.

**Review by mitruska — APPROVED** [review]
LGTM, but would be great to cover this change with corresponding tests.
The SDPA spec clarification can be handled as a follow up.

**Pull request merged**